File: 22.括号生成.go
/*
 * @lc app=leetcode.cn id=22 lang=golang
 *
 * [22] 括号生成
 */

// @lc code=start
func generateParenthesis(n int) []string {
	var paths []string
	var path []byte
	var dfs func(left, right int)

	dfs = func(left, right int) {
		if left == n && right == n {
			paths = append(paths, string(path))
			return
		}

		if left <= n {
			path = append(path, '(')
			dfs(left+1, right)
			path = path[:len(path)-1]
		}

		if right < left {
			path = append(path, ')')
			dfs(left, right+1)
			path = path[:len(path)-1]
		}
	}

	dfs(0, 0)

	return paths
}

// @lc code=end

File: 22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python
#
# [22] 括号生成
#

# @lc code=start
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        paths = []
        path = []
        def dfs(left, right):
            if left == n and right == n:
                paths.append("".join(path))
                return
            if left < n:
                path.append("(")
                dfs(left+1, right)
                path.pop()
            if right < left:
                path.append(")")
                dfs(left, right+1)
                path.pop()
        dfs(0, 0)
        return paths
# @lc code=end

